luaui/Widgets/api_resource_spot_builder.lua

```diff
@@ -142,122 +142,212 @@ local function IsSpotOccupied(spot)
 	return false
 end
 
+
+local function resourceSpotHasExistingExtractorCommand(x, z, builders)
+	for i=1, #builders do
+		local queue = Spring.GetCommandQueue(builders[i], 100)
+		for j=1, #queue do
+			local command = queue[j]
+			local id = -command.id
+			if(mexBuildings[id] or geoBuildings[id]) then
+				local dist = math.sqrt(distance2dSquared(x, z, command.params[1], command.params[3]))
+				if dist < Game_extractorRadius * 2 then
+					return true
+				end
+			end
+		end
+	end
+	return false
+end
+
+
+-- Naive best match, ignores special mexes (exploiter etc), just finds highest extraction amount
+local function getBestMexFromSelectedBuilders(units, constructorIds, extractors)
+	local bestExtraction = 0
+	local bestExtractor
+	for i = 1, #units do
+		-- only processes first mex option for each builder
+		local id = units[i]
+		local constructor = constructorIds[id]
+
+
+		if constructor then
+			local buildingID = -constructor.building[1]
+			local extractionAmount = extractors[buildingID]
+			if(extractionAmount > bestExtraction) then
+				bestExtraction = extractionAmount
+				bestExtractor = buildingID
+			end
+		end
+	end
+	return bestExtractor
+end
+
 -- Is there any better and allied mex at this location? (returns false if there is)
-local function NoAlliedMex(x, z, batchextracts)
-	local mexesatspot = Spring.GetUnitsInCylinder(x, z, Game_extractorRadius)
-	for i = 1, #mexesatspot do
-		local uid = mexesatspot[i]
-		if mexBuildings[spGetUnitDefID(uid)] and Spring.AreTeamsAllied(spGetMyTeamID(), Spring.GetUnitTeam(uid)) and mexBuildings[spGetUnitDefID(uid)] >= batchextracts then
-			return false
+local function canExtractorBeUpgraded(x, z, extractorId)
+
+	local newExtractor = UnitDefs[extractorId]
+	local newExtractorStrength = mexBuildings[extractorId] or geoBuildings[extractorId]
+	local newExtractorIsSpecial = newExtractor.stealth or #newExtractor.weapons > 0
+
+	local units = Spring.GetUnitsInCylinder(x, z, Game_extractorRadius)
+	for i = 1, #units do
+		local uid = units[i]
+		local uDefId = spGetUnitDefID(uid)
+		local currentExtractorStrength = mexBuildings[uDefId] or geoBuildings[uDefId]
+		if currentExtractorStrength then
+			if(newExtractorStrength > currentExtractorStrength) then
+				return true
+			end
+			if(newExtractorStrength == currentExtractorStrength and newExtractorIsSpecial) then
+				return true
+			end
 		end
 	end
-	return true
+
+	if #units == 0 then
+		return true
+	end
+
+	return false
 end
 
 
-local function BuildResourceExtractors(params, options, isGuard, justDraw, constructorIds, buildingIds, spots, checkDuplicateOrders)		-- when isGuard: needs to be a table of the unit pos: { x = ux, y = uy, z = uz }
+local function tablelength(T)
```

Comment:
Check if [`table.count`](https://github.com/Beyond-all-reason/Beyond-All-Reason/blob/master/common/tablefunctions.lua#L126-L132) is defined and use it instead

Reply:
Nice, hadn't found these before